File: server/solotrace/chords.py
from __future__ import annotations

import hashlib
import json
import uuid
from collections.abc import Callable
from functools import lru_cache
from pathlib import Path

import librosa
import numpy as np
from scipy.ndimage import uniform_filter1d

from .models import (
    ChordAlternative,
    ChordEvent,
    ChordQuality,
    ChordTrack,
    Project,
    SpelledPitch,
    TabDocument,
)
from .timing import audio_frame_to_score_tick

MODEL_REVISION = "fbd620e6a7617bbc82795b1f0c828a7721c213f4"
MODEL_SHA256 = "9a6570bf611cdc3f2c36286307af46fb94927fe7f6a2bc22a87c0ebf5f6c082e"
CONFIG_SHA256 = "1f26c11ebea51ec08f12e813eb213a729fa0ecc407ac7632dfdc7bad67e65aa4"
ENGINE_NAME = "ChordMini 2E1D ONNX"
CHUNK_SECONDS = 5 * 60
QUALITIES: tuple[ChordQuality, ...] = (
    "min",
    "maj",
    "dim",
    "aug",
    "min6",
    "maj6",
    "min7",
    "minmaj7",
    "maj7",
    "7",
    "dim7",
    "hdim7",
    "sus2",
    "sus4",
)

_RESOURCE_DIR = Path(__file__).resolve().parent / "resources" / "chordmini"
_MODEL_PATH = _RESOURCE_DIR / "chordnet.onnx"
_CONFIG_PATH = _RESOURCE_DIR / "config.json"


class ChordRecognitionCancelled(RuntimeError):
    pass


class ChordRecognitionUnavailable(RuntimeError):
    pass


def _sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as source:
        for block in iter(lambda: source.read(1024 * 1024), b""):
            digest.update(block)
    return digest.hexdigest()


@lru_cache(maxsize=1)
def model_config() -> dict[str, object]:
    if not _CONFIG_PATH.is_file() or _sha256(_CONFIG_PATH) != CONFIG_SHA256:
        raise ChordRecognitionUnavailable(
            "ChordMini vocabulary configuration is missing or damaged"
        )
    config = json.loads(_CONFIG_PATH.read_text())
    expected = {
        "seqLen": 108,
        "sampleRate": 22050,
        "hopLength": 2048,
        "nBins": 144,
        "binsPerOctave": 24,
        "smoothingKernel": 9,
        "numChords": 170,
    }
    if any(config.get(key) != value for key, value in expected.items()):
        raise ChordRecognitionUnavailable("ChordMini preprocessing configuration does not match")
    vocabulary = config.get("chordVocab")
    if not isinstance(vocabulary, list) or len(vocabulary) != 170:
        raise ChordRecognitionUnavailable("ChordMini vocabulary must contain 170 labels")
    return config


@lru_cache(maxsize=1)
def _model_status() -> tuple[bool, str]:
    try:
        model_config()
        if not _MODEL_PATH.is_file() or _sha256(_MODEL_PATH) != MODEL_SHA256:
            return False, "Pinned ChordMini model is missing or damaged"
        import onnxruntime  # noqa: F401
    except (ImportError, ChordRecognitionUnavailable) as error:
        return False, str(error) or "ONNX Runtime is unavailable"
    return True, "Pinned model verified for offline recognition"


def recognition_capability() -> dict[str, object]:
    available, detail = _model_status()
    return {
        "available": available,
        "engine": ENGINE_NAME,
        "modelRevision": MODEL_REVISION,
        "modelSha256": MODEL_SHA256,
        "detail": detail,
        "desktopOnly": True,
    }


@lru_cache(maxsize=1)
def _session():
    available, detail = _model_status()
    if not available:
        raise ChordRecognitionUnavailable(detail)
    import onnxruntime as ort

    return ort.InferenceSession(
        str(_MODEL_PATH),
        providers=["CPUExecutionProvider"],
    )


def _spelled_pitch(label: str) -> SpelledPitch:
    step = label[0]
    accidental = label[1:]
    return SpelledPitch(step=step, alter={"bb": -2, "b": -1, "": 0, "#": 1, "##": 2}[accidental])


def _label_parts(label: str) -> tuple[str, SpelledPitch | None, ChordQuality | None]:
    if label == "N":
        return "no-chord", None, None
    if label == "X":
        return "unknown", None, None
    root_text, separator, quality_text = label.partition(":")
    quality: ChordQuality = "maj" if not separator else quality_text  # type: ignore[assignment]
    if quality not in QUALITIES:
        raise ValueError(f"Unsupported ChordMini quality: {quality}")
    return "chord", _spelled_pitch(root_text), quality


def chord_symbol(event: ChordEvent) -> str:
    if event.kind == "no-chord":
        return "N.C."
    if event.kind == "unknown":
        return "X"
    assert event.root is not None and event.quality is not None
    accidental = {-2: "bb", -1: "b", 0: "", 1: "#", 2: "##"}[event.root.alter]
    suffix = {
        "maj": "",
        "min": "m",
        "dim": "dim",
        "aug": "aug",
        "min6": "m6",
        "maj6": "6",
        "min7": "m7",
        "minmaj7": "m(maj7)",
        "maj7": "maj7",
        "7": "7",
        "dim7": "dim7",
        "hdim7": "m7b5",
        "sus2": "sus2",
        "sus4": "sus4",
    }[event.quality]
    bass = ""
    if event.bass is not None:
        bass_accidental = {-2: "bb", -1: "b", 0: "", 1: "#", 2: "##"}[event.bass.alter]
        bass = f"/{event.bass.step}{bass_accidental}"
    return f"{event.root.step}{accidental}{suffix}{bass}"


def _softmax(logits: np.ndarray) -> np.ndarray:
    shifted = logits - np.max(logits, axis=-1, keepdims=True)
    exp = np.exp(shifted)
    return exp / np.sum(exp, axis=-1, keepdims=True)


def _infer_features(features: np.ndarray) -> np.ndarray:
    config = model_config()
    sequence_length = int(config["seqLen"])
    frame_count = features.shape[0]
    padded_count = ((frame_count + sequence_length - 1) // sequence_length) * sequence_length
    padded = np.pad(
        features.astype(np.float32, copy=False),
        ((0, padded_count - frame_count), (0, 0)),
        mode="constant",
    )
    windows = padded.reshape(-1, sequence_length, features.shape[1])
    logits = _session().run(["logits"], {"features": windows})[0]
    return logits.reshape(-1, logits.shape[-1])[:frame_count]


def _chunk_predictions(
    audio_path: Path,
    start_s: float,
    end_s: float,
    *,
    cancelled: Callable[[], bool],
) -> tuple[np.ndarray, np.ndarray]:
    config = model_config()
    sample_rate = int(config["sampleRate"])
    hop_length = int(config["hopLength"])
    sequence_length = int(config["seqLen"])
    overlap_s = sequence_length * hop_length / sample_rate
    frame_times: list[np.ndarray] = []
    frame_logits: list[np.ndarray] = []
    cursor = start_s
    while cursor < end_s - 1e-9:
        if cancelled():
            raise ChordRecognitionCancelled("Chord recognition cancelled")
        core_end = min(end_s, cursor + CHUNK_SECONDS)
        read_start = max(start_s, cursor - overlap_s)
        read_end = min(end_s, core_end + overlap_s)
        audio, _ = librosa.load(
            audio_path,
            sr=sample_rate,
            mono=True,
            offset=read_start,
            duration=read_end - read_start,
        )
        if cancelled():
            raise ChordRecognitionCancelled("Chord recognition cancelled")
        cqt = librosa.cqt(
            audio,
            sr=sample_rate,
            hop_length=hop_length,
            fmin=float(config["fmin"]),
            n_bins=int(config["nBins"]),
            bins_per_octave=int(config["binsPerOctave"]),
            norm=1,
            sparsity=0.01,
            window="hann",
            scale=True,
            pad_mode="constant",
        )
        features = np.log(np.abs(cqt).T + 1e-6).astype(np.float32)
        logits = _infer_features(features)
        times = read_start + librosa.frames_to_time(
            np.arange(features.shape[0]),
            sr=sample_rate,
            hop_length=hop_length,
        )
        keep = (times >= cursor - 1e-9) & (
            (times < core_end - 1e-9) if core_end < end_s else (times <= core_end + 1e-9)
        )
        frame_times.append(times[keep])
        frame_logits.append(logits[keep])
        cursor = core_end
    if not frame_times or not any(part.size for part in frame_times):
        raise RuntimeError("ChordMini produced no analyzable frames")
    times = np.concatenate(frame_times)
    logits = np.concatenate(frame_logits)
    order = np.argsort(times, kind="stable")
    times = times[order]
    logits = logits[order]
    unique = np.concatenate(([True], np.diff(times) > 1e-6))
    return times[unique], logits[unique]


def _event_from_segment(
    label_index: int,
    segment_logits: np.ndarray,
    onset_s: float,
    offset_s: float,
    tab: TabDocument,
) -> ChordEvent:
    vocabulary = model_config()["chordVocab"]
    assert isinstance(vocabulary, list)
    probabilities = _softmax(segment_logits)
    mean_scores = probabilities.mean(axis=0)
    top_indices = np.argsort(mean_scores)[::-1][:3]
    kind, root, quality = _label_parts(str(vocabulary[label_index]))
    alternatives = []
    for index in top_indices:
        alternative_kind, alternative_root, alternative_quality = _label_parts(
            str(vocabulary[int(index)])
        )
        alternatives.append(
            ChordAlternative(
                kind=alternative_kind,
                root=alternative_root,
                quality=alternative_quality,
                model_score=float(mean_scores[index]),
            )
        )
    onset_frame = round(onset_s * tab.sample_rate)
    end_frame = max(onset_frame + 1, round(offset_s * tab.sample_rate))
    score_tick = audio_frame_to_score_tick(onset_frame, tab.sync_anchors)
    end_tick = audio_frame_to_score_tick(end_frame, tab.sync_anchors)
    return ChordEvent(
        id=f"chord-{uuid.uuid4().hex[:12]}",
        onset_frame=onset_frame,
        end_frame=end_frame,
        audio_onset_s=onset_s,
        audio_offset_s=offset_s,
        score_tick=score_tick,
        duration_ticks=max(1, end_tick - score_tick),
        kind=kind,
        root=root,
        quality=quality,
        model_score=float(mean_scores[label_index]),
        alternatives=alternatives,
        provenance="detected",
        edited=False,
        reviewed=False,
    )


def recognize_chords(
    audio_path: Path,
    start_s: float,
    end_s: float,
    tab: TabDocument,
    *,
    cancelled: Callable[[], bool] = lambda: False,
) -> ChordTrack:
    available, detail = _model_status()
    if not available:
        raise ChordRecognitionUnavailable(detail)
    times, logits = _chunk_predictions(
        audio_path,
        start_s,
        end_s,
        cancelled=cancelled,
    )
    smoothed = uniform_filter1d(
        logits,
        size=int(model_config()["smoothingKernel"]),
        axis=0,
        mode="nearest",
    )
    labels = np.argmax(smoothed, axis=1)
    events: list[ChordEvent] = []
    segment_start = 0
    for index in range(1, len(labels) + 1):
        if index < len(labels) and labels[index] == labels[segment_start]:
            continue
        onset_s = start_s if segment_start == 0 else float(times[segment_start])
        offset_s = end_s if index == len(labels) else float(times[index])
        if offset_s > onset_s:
            events.append(
                _event_from_segment(
                    int(labels[segment_start]),
                    smoothed[segment_start:index],
                    onset_s,
                    offset_s,
                    tab,
                )
            )
        segment_start = index
    return ChordTrack(
        engine=ENGINE_NAME,
        model_revision=MODEL_REVISION,
        model_sha256=MODEL_SHA256,
        analyzed_start_s=start_s,
        analyzed_end_s=end_s,
        events=events,
    )


def normalize_edited_chords(
    project: Project,
    submitted: ChordTrack,
) -> ChordTrack:
    events = sorted(submitted.events, key=lambda event: (event.audio_onset_s, event.id))
    ids = [event.id for event in events]
    if len(ids) != len(set(ids)):
        raise ValueError("Chord ids must be unique")
    if not events:
        return submitted.model_copy(
            update={
                "engine": submitted.engine or "manual",
                "events": [],
            }
        )
    range_start = submitted.analyzed_start_s
    range_end = submitted.analyzed_end_s
    if range_start is None or range_end is None:
        raise ValueError("A non-empty chord track needs an analyzed range")
    if range_start < project.passage.start_s or range_end > min(
        project.passage.end_s,
        project.duration_s,
    ):
        raise ValueError("Chord range must stay inside the transcription range")
    tolerance = 1e-4
    if abs(events[0].audio_onset_s - range_start) > tolerance:
        raise ValueError("Chord spans must start at the analyzed range")
    if abs(events[-1].audio_offset_s - range_end) > tolerance:
        raise ValueError("Chord spans must end at the analyzed range")
    normalized: list[ChordEvent] = []
    for index, event in enumerate(events):
        if index and abs(events[index - 1].audio_offset_s - event.audio_onset_s) > tolerance:
            raise ValueError("Chord spans must be contiguous and cannot overlap")
        onset_s = range_start if index == 0 else normalized[-1].audio_offset_s
        offset_s = range_end if index == len(events) - 1 else event.audio_offset_s
        if offset_s <= onset_s:
            raise ValueError("Chord spans must have positive duration")
        onset_frame = round(onset_s * project.tab.sample_rate)
        end_frame = max(onset_frame + 1, round(offset_s * project.tab.sample_rate))
        score_tick = audio_frame_to_score_tick(onset_frame, project.tab.sync_anchors)
        end_tick = audio_frame_to_score_tick(end_frame, project.tab.sync_anchors)
        normalized.append(
            event.model_copy(
                update={
                    "onset_frame": onset_frame,
                    "end_frame": end_frame,
                    "audio_onset_s": onset_s,
                    "audio_offset_s": offset_s,
                    "score_tick": score_tick,
                    "duration_ticks": max(1, end_tick - score_tick),
                }
            )
        )
    return submitted.model_copy(update={"events": normalized})

File: packages/editor/src/chords.ts
import type {
  ChordEvent,
  ChordKind,
  ChordQuality,
  ChordTrack,
  SpelledPitch,
  SyncAnchor,
  TabDocument,
} from './types'

export const CHORD_QUALITIES: readonly ChordQuality[] = Object.freeze([
  'min',
  'maj',
  'dim',
  'aug',
  'min6',
  'maj6',
  'min7',
  'minmaj7',
  'maj7',
  '7',
  'dim7',
  'hdim7',
  'sus2',
  'sus4',
])

export function emptyChordTrack(): ChordTrack {
  return {
    engine: 'manual',
    model_revision: null,
    model_sha256: null,
    analyzed_start_s: null,
    analyzed_end_s: null,
    events: [],
  }
}

function accidental(pitch: SpelledPitch): string {
  return ({ [-2]: 'bb', [-1]: 'b', [0]: '', [1]: '#', [2]: '##' } as Record<number, string>)[
    pitch.alter
  ] ?? ''
}

const QUALITY_SUFFIX: Record<ChordQuality, string> = {
  min: 'm',
  maj: '',
  dim: 'dim',
  aug: 'aug',
  min6: 'm6',
  maj6: '6',
  min7: 'm7',
  minmaj7: 'm(maj7)',
  maj7: 'maj7',
  '7': '7',
  dim7: 'dim7',
  hdim7: 'm7b5',
  sus2: 'sus2',
  sus4: 'sus4',
}

const SUFFIX_QUALITY = new Map<string, ChordQuality>([
  ['', 'maj'],
  ['maj', 'maj'],
  ['m', 'min'],
  ['-', 'min'],
  ['min', 'min'],
  ['dim', 'dim'],
  ['°', 'dim'],
  ['aug', 'aug'],
  ['+', 'aug'],
  ['m6', 'min6'],
  ['min6', 'min6'],
  ['6', 'maj6'],
  ['maj6', 'maj6'],
  ['m7', 'min7'],
  ['min7', 'min7'],
  ['m(maj7)', 'minmaj7'],
  ['minmaj7', 'minmaj7'],
  ['mmaj7', 'minmaj7'],
  ['maj7', 'maj7'],
  ['7', '7'],
  ['dim7', 'dim7'],
  ['°7', 'dim7'],
  ['hdim7', 'hdim7'],
  ['m7b5', 'hdim7'],
  ['ø7', 'hdim7'],
  ['sus2', 'sus2'],
  ['sus4', 'sus4'],
  ['sus', 'sus4'],
])

export function formatChordSymbol(event: Pick<ChordEvent, 'kind' | 'root' | 'quality' | 'bass'>) {
  if (event.kind === 'no-chord') return 'N.C.'
  if (event.kind === 'unknown') return 'X'
  if (!event.root || !event.quality) return 'X'
  const bass = event.bass ? `/${event.bass.step}${accidental(event.bass)}` : ''
  return `${event.root.step}${accidental(event.root)}${QUALITY_SUFFIX[event.quality]}${bass}`
}

function parsePitch(text: string): SpelledPitch | null {
  const match = /^([A-Ga-g])(bb|##|b|#)?$/.exec(text.trim())
  if (!match) return null
  return {
    step: match[1].toUpperCase() as SpelledPitch['step'],
    alter: ({ bb: -2, b: -1, '#': 1, '##': 2 } as Record<string, number>)[match[2] ?? ''] ?? 0,
  }
}

export type ParsedChordSymbol = {
  kind: ChordKind
  root: SpelledPitch | null
  quality: ChordQuality | null
  bass: SpelledPitch | null
}

export function parseChordSymbol(input: string): ParsedChordSymbol {
  const symbol = input.trim()
  if (/^(?:N\.?C\.?|no[- ]?chord)$/i.test(symbol)) {
    return { kind: 'no-chord', root: null, quality: null, bass: null }
  }
  if (/^(?:X|\?|unknown)$/i.test(symbol)) {
    return { kind: 'unknown', root: null, quality: null, bass: null }
  }
  const [body, bassText, ...extra] = symbol.split('/')
  if (extra.length) throw new Error('Use one optional slash bass, such as C/E')
  const rootMatch = /^([A-Ga-g](?:bb|##|b|#)?)(.*)$/.exec(body)
  if (!rootMatch) throw new Error('Enter a chord such as C, Cm7, Dbmaj7, or N.C.')
  const root = parsePitch(rootMatch[1])
  const quality = SUFFIX_QUALITY.get(rootMatch[2].toLowerCase())
  const bass = bassText === undefined ? null : parsePitch(bassText)
  if (!root || !quality || (bassText !== undefined && !bass)) {
    throw new Error('This chord spelling or quality is not supported')
  }
  return { kind: 'chord', root, quality, bass }
}

function audioFrameToScoreTick(audioFrame: number, anchors: readonly SyncAnchor[]): number {
  if (!anchors.length) return 0
  const ordered = [...anchors].sort((left, right) => left.audio_frame - right.audio_frame)
  if (audioFrame <= ordered[0].audio_frame) return ordered[0].score_tick
  for (let index = 0; index < ordered.length - 1; index += 1) {
    const left = ordered[index]
    const right = ordered[index + 1]
    if (audioFrame <= right.audio_frame) {
      const span = right.audio_frame - left.audio_frame
      if (!span) return right.score_tick
      return Math.round(
        left.score_tick +
          ((audioFrame - left.audio_frame) / span) * (right.score_tick - left.score_tick),
      )
    }
  }
  const last = ordered.at(-1)!
  const previous = ordered.at(-2) ?? last
  const span = last.audio_frame - previous.audio_frame
  if (!span) return last.score_tick
  return Math.max(
    0,
    Math.round(
      last.score_tick +
        (audioFrame - last.audio_frame) *
          ((last.score_tick - previous.score_tick) / span),
    ),
  )
}

export function normalizeChordTrack(track: ChordTrack, tab: TabDocument): ChordTrack {
  const events = [...track.events].sort(
    (left, right) => left.audio_onset_s - right.audio_onset_s || left.id.localeCompare(right.id),
  )
  if (!events.length) return { ...track, events: [] }
  if (track.analyzed_start_s === null || track.analyzed_end_s === null) {
    throw new Error('A non-empty chord track needs an analyzed range')
  }
  const ids = new Set(events.map((event) => event.id))
  if (ids.size !== events.length) throw new Error('Chord IDs must be unique')
  if (Math.abs(events[0].audio_onset_s - track.analyzed_start_s) > 0.0001) {
    throw new Error('Chords must start at the analyzed range')
  }
  if (Math.abs(events.at(-1)!.audio_offset_s - track.analyzed_end_s) > 0.0001) {
    throw new Error('Chords must end at the analyzed range')
  }
  return {
    ...track,
    events: events.map((event, index) => {
      const onset =
        index === 0 ? track.analyzed_start_s! : events[index - 1].audio_offset_s
      const offset =
        index === events.length - 1 ? track.analyzed_end_s! : event.audio_offset_s
      if (offset <= onset) throw new Error('Chord spans must have positive duration')
      if (
        index &&
        Math.abs(events[index - 1].audio_offset_s - event.audio_onset_s) > 0.0001
      ) {
        throw new Error('Chord spans must be contiguous and cannot overlap')
      }
      const onsetFrame = Math.round(onset * tab.sample_rate)
      const endFrame = Math.max(onsetFrame + 1, Math.round(offset * tab.sample_rate))
      const scoreTick = audioFrameToScoreTick(onsetFrame, tab.sync_anchors)
      const endTick = audioFrameToScoreTick(endFrame, tab.sync_anchors)
      return {
        ...event,
        onset_frame: onsetFrame,
        end_frame: endFrame,
        audio_onset_s: onset,
        audio_offset_s: offset,
        score_tick: scoreTick,
        duration_ticks: Math.max(1, endTick - scoreTick),
      }
    }),
  }
}

export function replaceChordSymbol(
  track: ChordTrack,
  chordId: string,
  symbol: string,
): ChordTrack {
  const parsed = parseChordSymbol(symbol)
  return {
    ...track,
    events: track.events.map((event) =>
      event.id === chordId
        ? {
            ...event,
            ...parsed,
            provenance: 'manual',
            edited: true,
          }
        : event,
    ),
  }
}

export function setChordReviewed(
  track: ChordTrack,
  chordId: string,
  reviewed: boolean,
): ChordTrack {
  return {
    ...track,
    events: track.events.map((event) =>
      event.id === chordId ? { ...event, reviewed } : event,
    ),
  }
}

export function splitChord(
  track: ChordTrack,
  chordId: string,
  atSeconds: number,
  newId: string,
): ChordTrack {
  const index = track.events.findIndex((event) => event.id === chordId)
  if (index < 0) throw new Error('Chord not found')
  const event = track.events[index]
  if (atSeconds <= event.audio_onset_s || atSeconds >= event.audio_offset_s) {
    throw new Error('Split point must be inside the chord')
  }
  const left = { ...event, audio_offset_s: atSeconds, edited: true }
  const right = {
    ...event,
    id: newId,
    audio_onset_s: atSeconds,
    provenance: 'manual' as const,
    edited: true,
    reviewed: false,
  }
  return {
    ...track,
    events: [...track.events.slice(0, index), left, right, ...track.events.slice(index + 1)],
  }
}

export function mergeChord(
  track: ChordTrack,
  chordId: string,
  direction: 'left' | 'right',
): ChordTrack {
  const index = track.events.findIndex((event) => event.id === chordId)
  const neighborIndex = direction === 'left' ? index - 1 : index + 1
  if (index < 0 || neighborIndex < 0 || neighborIndex >= track.events.length) {
    throw new Error(`No chord to merge on the ${direction}`)
  }
  const first = Math.min(index, neighborIndex)
  const second = Math.max(index, neighborIndex)
  const selected = track.events[index]
  const merged = {
    ...selected,
    audio_onset_s: track.events[first].audio_onset_s,
    audio_offset_s: track.events[second].audio_offset_s,
    provenance: 'manual' as const,
    edited: true,
  }
  return {
    ...track,
    events: [...track.events.slice(0, first), merged, ...track.events.slice(second + 1)],
  }
}

export function moveChordBoundary(
  track: ChordTrack,
  leftChordId: string,
  atSeconds: number,
): ChordTrack {
  const index = track.events.findIndex((event) => event.id === leftChordId)
  if (index < 0 || index === track.events.length - 1) throw new Error('Boundary not found')
  const left = track.events[index]
  const right = track.events[index + 1]
  if (atSeconds <= left.audio_onset_s || atSeconds >= right.audio_offset_s) {
    throw new Error('Boundary must stay between the surrounding chord edges')
  }
  return {
    ...track,
    events: track.events.map((event, eventIndex) => {
      if (eventIndex === index) return { ...event, audio_offset_s: atSeconds, edited: true }
      if (eventIndex === index + 1) {
        return { ...event, audio_onset_s: atSeconds, edited: true }
      }
      return event
    }),
  }
}

export function deleteChordToUnknown(track: ChordTrack, chordId: string): ChordTrack {
  return {
    ...track,
    events: track.events.map((event) =>
      event.id === chordId
        ? {
            ...event,
            kind: 'unknown',
            root: null,
            quality: null,
            bass: null,
            provenance: 'manual',
            edited: true,
            reviewed: false,
          }
        : event,
    ),
  }
}
